libkbfs/folder_branch_ops.go

```diff
@@ -791,6 +792,236 @@ func (fbo *folderBranchOps) startMonitorChat(tlfName tlf.CanonicalName) {
+		if !ok {
+			return keybase1.FolderSyncConfig{}, errors.Errorf(
+				"Partial sync block is not a file block, but %T", b)
+		}
+	} else {
+		block = NewFileBlock().(*FileBlock)
+		err = assembleBlock(
+			ctx, fbo.config.keyGetter(), fbo.config.Codec(),
+			fbo.config.Crypto(), kmd, config.Paths.Ptr, block,
+			config.Paths.Buf, config.Paths.ServerHalf)
+		if err != nil {
+			return keybase1.FolderSyncConfig{}, err
+		}
+	}
+
+	paths, err := syncPathListFromBlock(fbo.config.Codec(), block)
+	if err != nil {
+		return keybase1.FolderSyncConfig{}, err
+	}
+	ret.Paths = paths.Paths
+	return ret, nil
+}
+
+func (fbo *folderBranchOps) getProtocolSyncConfigUnlocked(
+	ctx context.Context, lState *lockState, kmd KeyMetadata) (
+	ret keybase1.FolderSyncConfig, err error) {
+	fbo.syncLock.RLock(lState)
+	defer fbo.syncLock.RUnlock(lState)
+	return fbo.getProtocolSyncConfig(ctx, lState, kmd)
+}
+
+func (fbo *folderBranchOps) syncOneNode(
+	ctx context.Context, node Node, rmd ImmutableRootMetadata,
+	action BlockRequestAction) (BlockPointer, error) {
+	nodePath := fbo.nodeCache.PathFromNode(node)
+	var b Block
+	if node.EntryType() == Dir {
+		b = NewDirBlock()
+	} else {
+		b = NewFileBlock()
+	}
+	ptr := nodePath.tailPointer()
+	ch := fbo.config.BlockOps().BlockRetriever().Request(
+		ctx, defaultOnDemandRequestPriority-1, rmd,
+		ptr, b, TransientEntry, action)
+	select {
+	case err := <-ch:
+		if err != nil {
+			return zeroPtr, err
+		}
+		return ptr, nil
+	case <-ctx.Done():
+		return zeroPtr, ctx.Err()
+	}
+}
+
+func (fbo *folderBranchOps) doPartialSync(
+	ctx context.Context, syncConfig keybase1.FolderSyncConfig,
+	latestMerged ImmutableRootMetadata) (err error) {
+	fbo.log.CDebugf(
+		ctx, "Starting partial sync at revision %d", latestMerged.Revision())
+	defer func() {
+		fbo.deferLog.CDebugf(ctx, "Partial sync done: %+v", err)
+	}()
+
+	if syncConfig.Mode != keybase1.FolderSyncMode_PARTIAL {
+		return errors.Errorf(
+			"Bad mode passed to partial sync: %+v", syncConfig.Mode)
+	}
+
+	rootNode, _, _, err := fbo.getRootNode(ctx)
+	if err != nil {
+		return err
+	}
+	// Use `PrefetchTail` for directories, to make sure that any child
+	// blocks in the directory itself get prefetched.
+	_, err = fbo.syncOneNode(
+		ctx, rootNode, latestMerged, BlockRequestPrefetchTailWithSync)
+	if err != nil {
+		return err
+	}
+
+	chs := make(map[string]<-chan struct{}, len(syncConfig.Paths))
+	// Look up and solo-sync each lead-up component of the path.
+pathLoop:
+	for _, p := range syncConfig.Paths {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		fbo.log.CDebugf(ctx, "Partially-syncing %s", p)
+
+		parentPath, syncedElem := stdpath.Split(p)
+		parents := strings.Split(strings.TrimSuffix(parentPath, "/"), "/")
+		currNode := rootNode
+		for _, parent := range parents {
+			if len(parent) == 0 {
+				continue
+			}
+			// TODO: parallelize the parent fetches and lookups.
+			currNode, _, err = fbo.Lookup(ctx, currNode, parent)
```

Comment:
Won't this already trigger a prefetch? I guess it's fine, though, since you escalate the prefetch action with the explicit `syncOneNode` call later.